### PR TITLE
Sanitization of Algorithm cancellation mechanics.

### LINKED ISSUE #15145: Algorithm cancellation flag mechanism looks suspicious when threaded.
The fields `m_cancel` and `m_parallelException` of `Algorithm` are read and written to from different threads. However, they are neither atomic, nor do they seem to be consistently protected by a mutex (in particular in the `PARALLEL` macros in `MultiThreaded.h`). This can potentially make the cancel/exception mechanism useless (i.e., other threads will not abort).

To illustrate the issue, the following works when compiled with `g++ -O0 -fopenmp` but deadlocks with `-O1` or higher:

```cpp
#include <omp.h>

int main() {
  bool cancel = false;
#pragma omp parallel num_threads(10)
  {
    int id = omp_get_thread_num();
    if(id == 0) {
      for(size_t i=0; i<10000000; i++) {}
      cancel = true;
    } else {
      while (!cancel) {
      }
    }
  }
}
```

### PR DESCRIPTION
**Pull request by SimonHeybrock**
Fixes #15145.

Code review:
Someone with good understanding of threading. I am not 100% sure I fully understood the (intent of) existing mechanism (it seemed flawed in some points), but I believe this makes more sense and is safer.

Testing:
See if algorithms can still be canceled as they should (in particular threaded algorithms).

No changes for release notes.

### CODE REVIEW

**Comment by martyngigg** [comment]
@SimonHeybrock The branch has developed some merge conflicts with master.

File: Framework/API/src/Algorithm.cpp [hunk 1]
```diff
@@ -1524,7 +1517,6 @@ void Algorithm::cancel() {
  * and check if the algorithm has requested that it be cancelled.
  */
 void Algorithm::interruption_point() {
-  Poco::FastMutex::ScopedLock _lock(m_mutex);
```
Comment (martyngigg) [inline Framework/API/src/Algorithm.cpp]: This certainly does seem completely unnecessary given the `IF_NOT_PARALLEL` below. I'm not sure why that is required either but I think that is a question for another day.

**Comment by martyngigg** [comment]
Agreed that using `atomic`s here makes much more sense.

File: Framework/API/src/Algorithm.cpp (hunk 1 above)
Comment (SimonHeybrock) [inline Framework/API/src/Algorithm.cpp]: Were you recommending to not touch it?

Comment (martyngigg) [inline Framework/API/src/Algorithm.cpp]: Yes, I think it would be outside of the scope of these changes.

Comment (SimonHeybrock) [inline Framework/API/src/Algorithm.cpp]: Yes, I guess the follow-up question is why `interruption_point()` is called in `PARALLEL_CHECK_INTERUPT_REGION`.

See #15213.

**Comment by martyngigg** [comment]
Functionally this all appears to work as expected.

**Pull request merged**